### DISCUSSION TITLE
Chumsky and Ariadne spans: Unicode chars or UTF8/bytes?

### DISCUSSION THREAD

**Discussion by ebastien** [zesterer/chumsky#493] [Q&A] [answered]
Looking at the code in Chumsky 1.0.0-alpha and Ariadne 0.3.0, I get the impression that default Chumsky's spans on `&str` inputs (`StrInput`) are counting UTF8/bytes (i.e. `str.len()`) but Ariadne's spans are counting Unicode characters (i.e. `str.chars().count()`).
Is that a correct assumption?

**Comment by ebastien** [comment]
```rust
use ariadne::{Label, Report, ReportKind, Source};
use chumsky::prelude::*;

fn main() {
    let input = "h😉hhh".to_owned();
    let p = any::<_, extra::Err<Simple<char>>>()
        .map_with_span(|c, s| (c, s))
        .repeated()
        .collect::<Vec<_>>();

    let r = p.parse(input.as_str()).into_output().unwrap();

    let mut builder = Report::build(ReportKind::Error, "input", 0).with_message("error");
    for (c, s) in r {
        builder.add_label(Label::new(("input", s.into_range())).with_message(format!("'{}'", c)));
    }
    builder
        .finish()
        .print(("input", Source::from(input)))
        .unwrap();
}
```

```
Error: error
   ╭─[input:1:1]
   │
 1 │ h😉hhh
   │ ┬───┬─  
   │ ╰─────── 'h'
   │     │   
   │     ╰─── '😉'
───╯
```

Are there best practices to nicely bridge Chumsky and Ariadne when dealing with Unicode?
I implemented a Stream input with a custom Span based on Unicode chars but I then keep fighting the combinators in Chumsky 1.0.0-alpha (e.g. `map_slice`).

**Comment by zesterer** [comment, marked as answer]
This is a correct assumption. Ariadne using character offsets instead of byte offsets was a misstep, and one I intend to correct at some point in the future. If you want to bridge the two, you can either:

1) Use character offsets in chumsky (`Stream` should allow this without too much difficulty)

2) Use `str::char_indices` to linearly search through the string to convert byte offsets to char offsets.
